Extensions/DialogueTree/dialoguetools.js

```diff
@@ -0,0 +1,583 @@
+ * @param {string} title The title of the branch to check for.
+ * Leaving this empty will check if the current branch title has been visited in the past.
+ */
+gdjs.dialogueTree.branchTitleHasBeenVisited = function(title) {
+  if (!title) title = this.dialogueBranchTitle;
+  return (
+    Object.keys(this.runner.visited).includes(title) &&
+    this.runner.visited[title]
+  );
+};
+
+/**
+ * use this to get the entire unparsed text of the current Dialogue Branch
+ */
+gdjs.dialogueTree.getBranchText = function() {
+  if (this.dialogueIsRunning) {
+    return this.dialogueBranchBody;
+  }
+  return "";
+};
+
+/**
+ * use this to get the value of a variable that was created by the Dialogue parses.
+ * @param {string} key The name of the variable you want to get the value of
+ */
+gdjs.dialogueTree.getVariable = function(key) {
+  if (this.dialogueIsRunning && key in this.runner.variables.data) {
+    return this.runner.variables.data[key];
+  }
+  return "";
+};
+
+/**
+ * use this to check if a specific variable created by the Dialogue parses exists and is equal to a specific value.
+ * @param {string} key The name of the variable you want to check the value of
+ * @param {string} value The value you want to check against
+ */
+gdjs.dialogueTree.compareVariable = function(key, value) {
+  if (this.dialogueIsRunning && key in this.runner.variables.data) {
+    return this.runner.variables.data[key].toString() === value;
+  }
+  return false;
+};
+
+/**
+ * Use this to Set a specific variable created by the Dialogue parser to a specific value.
+ * @param {string} key The name of the variable you want to set the value of
+ * @param {string} value The value you want to set
+ */
+gdjs.dialogueTree.setVariable = function(key, value) {
+  if (this.dialogueIsRunning && this.runner.variables.data) {
+    this.runner.variables.data[key] = value;
+  }
+};
+
+/**
+ * Use this to store the current State of the Dialogue Parser in a specified variable.
+ * Use this to implement persistence in dialogue through your game's Load/Save function.
+ * That way you can later load all the dialogue choices the player has made.
+ * @param {gdjs.Variable} storeVar The variable where to store the State
+ */
+gdjs.dialogueTree.saveState = function(storeVar) {
+  if (!this.dialogueIsRunning) return;
+  const dialogueState = {
+    variables: gdjs.dialogueTree.runner.variables.data,
+    visited: gdjs.dialogueTree.runner.visited
+  };
+  gdjs.evtTools.network._objectToVariable(dialogueState, storeVar);
+};
+
+/**
+ * Use this to load the current State of the Dialogue Parser from a specified variable.
+ * Use this to implement persistence in dialogue through your game's Load/Save function.
+ * That way you can later load all the dialogue choices the player has made.
+ * @param {gdjs.Variable} storeVar The variable where to load the State from.
+ */
+gdjs.dialogueTree.loadState = function(storeVar) {
+  if (!this.dialogueIsRunning) return;
```

Comment:
Is this condition valid? Can't we load a state if the dialogue is not running?

Reply:
good catch! This shouldn't be there. The developer should be able to load/save state before running any dialogue